packages/react-reconciler/src/ReactFiberCommitWork.js

```diff
@@ -339,15 +336,21 @@ function commitHookEffectList(
     const firstEffect = lastEffect.next;
     let effect = firstEffect;
     do {
-      if ((effect.tag & unmountTag) !== NoHookEffect) {
+      if (
+        (effect.tag & HookHasEffect) !== NoHookEffect &&
+        (effect.tag & unmountTag) !== NoHookEffect
+      ) {
         // Unmount
         const destroy = effect.destroy;
         effect.destroy = undefined;
         if (destroy !== undefined) {
           destroy();
         }
       }
-      if ((effect.tag & mountTag) !== NoHookEffect) {
+      if (
+        (effect.tag & HookHasEffect) !== NoHookEffect &&
+        (effect.tag & mountTag) !== NoHookEffect
```

Comment:
Nit: You can save a check here if you assume `mountTag` and `unmountTag` already include `HookHasEffect`:

```js
(effect.tag & mountTag) === mountTag
```

Reply:
Hmm! But they specifically don't.

We use the tag to specify the *type* of effect even if there's no work currently required by the hook. This lets us properly differentiate between passive and layout effects in the unmount case. (`NoHookEffect` is used to signal that there's *also* work for that specific hook during the current commit.)

Reply:
Ah, I see your comment below and I understand what you meant now. :smile: Sure I can make that change!

Reply:
I think pre-combing would cause false positives.

This would fail for a layout tag:
```js
let tag = Layout | HasEffect;

tag & HasEffect && tag & Passive // false
tag & HasEffect | Passive // true
```

Reply:
Please excuse my previous comment. I gotcha.